### PR TITLE
148 - ids-pager

### LINKED ISSUE #148: IdsPager: Add new web component
Add the standalone pager component ONLY. See https://main-enterprise.demo.design.infor.com/components/pager/example-standalone.html

The idea is to have a flexible slot so they can put in what pager items they want between:
- next/prev/left/right button
- page selector menu button
- small page selector menu button
- page input https://main-enterprise.demo.design.infor.com/components/pager/example-datagrid.html
- page numbers https://main-enterprise.demo.design.infor.com/components/pager/example-listview.html

then control the states / disabled ect

**Phase 1 Features (MVP)**
- Standalone pager
- All supported pager items can be put in as explained above
- Fires events will internally storing a page state object

**Future Features (Not Now)**
- integration with other components
- all old pager logic will be deprecated it will just be a "dumb" pager that holds the items and fires events

**Suggested Internal Markup**
```js
<ids-message>
  <div slot="title"></div>
  <div slot="message"></div>
  <div slot="buttons"></div>
</ids-message>
```

**Suggested Dependencies**
- ids-tooltip (on buttons)
- ids-text
- ids-button
- ids-input (page input)
- ids-menu-button (page selector)
- ids-hyperlink (numbers)

**Settings**
NA

**Design**
- https://www.figma.com/file/ri2Knf3KchdfdzRAeds0Ab/IDS-Mobility-v4.6?node-id=1%3A5740

**Events**
- pagechanged

**Other**
Pass the following checklist ordered in priority (some may be left for later) https://github.com/infor-design/enterprise-wc/blob/main/doc/CHECKLIST.md

### PR DESCRIPTION
**Pull request by rob2d**
**Explain the details for making this change. What existing problem does the pull request solve?**

Adds `ids-pager` (Phase 1) which includes:
- next/prev/left/right button
- page input https://main-enterprise.demo.design.infor.com/components/pager/example-datagrid.html
- page numbers https://main-enterprise.demo.design.infor.com/components/pager/example-listview.html
- disabled state (either via overall component, navigation, and user-per-button)

Also adds a `square` attribute to `ids-button` to support `ids-pager-button` having non-rounded corners.

Due to the scope and complexity of this one, the `page-size` menu isn't currently in this push but should be easier to faciliate on revisit as aligning/inserting content in right or left margin exists (can view this at `localhost:4300/ids-pager/sandbox`)

**Related github/jira issue (required)**:
Closes #148 

**Steps necessary to review your pull request (required)**:

- visit http://localhost:4300/ids-pager
- check the functionality listed above

**Included in this Pull Request**:
- [X] An e2e or functional test for the bug or feature.
- [x] A note to the change log. (TODO)

### CODE REVIEW

**Review by tmcconechy — COMMENTED** [review]
This is looking VERY nice. Thanks @rob2d just some minor thoughts.
Also in terms of functionality/UI. Can you center the text in the input?

File: app/ids-pager/example.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
+import IdsLayoutGrid from '../../src/ids-layout-grid/ids-layout-grid';
+import IdsLayoutGridCell from '../../src/ids-layout-grid/ids-layout-grid-cell';
+import IdsText from '../../src/ids-text/ids-text';
+import IdsButton from '../../src/ids-button/ids-button';
+
+const TemplateHTML = {
```
[the remark this answers is not in the document]
Comment (rob2d) [reply app/ids-pager/example.js:6]: thanks for the feedback on this one -- I was considering centering the text but I just wanted to try to simulate the older version. There was another issue where the input widths should grow a bit also when there's > 3 digits but it wasn't an edge case before so maybe I can try to address that while touching this via a class.

I think that also makes sense, but I had it as toggle-able to also test it (had discovered a few edge cases when rendering the different numberlist inside of that). Now that it works I can sort that out, but the con is due to some quirks in JSDOM/Jest its hard to catch these edge cases in our functional tests.

File: src/ids-pager/README.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+- Content can be laid out for additional buttons/interactions/etc other than the `ids-pager-button`, but the standard navigation and user input components should be used wherever possible.
+- the page count on the pager can be accessed on the `ids-pager` element at any time via the `pageCount` property via JS for example:
+```js
+  const idsElement = document.querySelector('ids-pager');
+  const pageCount = idsElement.pageCount;
+```
+
+## Accessiblity
+
+TODO
```
[the remark this answers is not in the document]
Comment (rob2d) [reply src/ids-pager/README.md:161]: Great point completely forgot that keyboard consideration is one of the #1 things, will do shortly thanks.

File: app/ids-pager/example.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (tmcconechy) [reply app/ids-pager/example.js:6]: Screen Shot 2021-06-24 at 12 12 07 PM

FYI it was centered (and deepak did add an option for this)

File: src/ids-pager/ids-pager-button.d.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import { IdsElement } from '../ids-base';
+
+export default class IdsPagerButton extends IdsElement {
+  /** 1-based page number displayed; managed by parent ids-pager */
+  pageNumber: number;
+
+  /**
```
[the remark this answers is not in the document]
Comment (rob2d) [reply src/ids-pager/ids-pager-button.d.ts:7]: makes sense -- thanks for pointing that out.

File: src/ids-pager/ids-pager-button.js [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+
+  /**
+   * @param {boolean|string} value designates this as a nav to last page button
+   */
+  set last(value) {
+    this.#setTypeAttribute(attributes.LAST, value);
+  }
+
+  /**
+   * @param {boolean|string} value designates this as a nav to next page button
```
[the remark this answers is not in the document]
Comment (rob2d) [reply src/ids-pager/ids-pager-button.js:96]: just resolving this one based on the convo as the IDE picked up (putting a comment in case anyone else hadn't known noticed which is completely understandable)

File: src/ids-pager/ids-pager-input.js [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+    // when leaving user focus, input should adjust itself
+    // to the page number already provided by the pager
+
+    this.onEvent('blur', this.input, /* istanbul ignore next */ () => {
+      if (this.input.value !== `${this.pageNumber}`) {
+        this.input.value = this.pageNumber;
+      }
+    });
+
+    // istanbul ignore next
```
[the remark this answers is not in the document]
Comment (rob2d) [reply src/ids-pager/ids-pager-input.js:86]: you may be right here -- I remember mentioning having an issue due to istanbul not picking things up and totally could just be my mistake. Will look into it along the way of these other things to see if that was an error/not useless for coverage.

[the remark this answers is not in the document]
Comment (rob2d) [reply src/ids-pager/ids-pager-input.js:86]: just confirmed that apparently yes, it does (!)

File: app/ids-pager/example.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (rob2d) [reply app/ids-pager/example.js:6]: sorry about that -- in that case I am not sure how the confusion happened there but I'll address it.

**Comment by rob2d** [comment]
@tmcconechy just a heads up I added a markdown rule to ignore `ordered-list-marker-value`since this was creating some false errors in 2 places. It seemed to be a little confusing reading some things where it would reset from step 3 to 1, but if you think the pros outweigh the cons I then I can remove it and leave the two instances where numbers reset to 1 there.

**Comment by tmcconechy** [comment]
> @tmcconechy just a heads up I added a markdown rule to ignore `ordered-list-marker-value`since this was creating some false errors in 2 places. It seemed to be a little confusing reading some things where it would reset from step 3 to 1, but if you think the pros outweigh the cons I then I can remove it and leave the two instances where numbers reset to 1 there.

Ok , i guess thats ok. But i think numbering with 1. is a lot easier as you can add more without renumbering the list but no big deal to make it optional

**Review by tmcconechy — APPROVED** [review]

**Review by EdwardCoyle — APPROVED** [review]
@rob2d Looks really nice!  Well done!
